### ISSUE TITLE
mismatched types error when  returning impl Type from a function, and using a feature

### ISSUE THREAD

**Issue by wiiznokes** [rust-lang/rust-clippy#12499]
### Description

This code compile correctly, but generate a `clippy` error:

```rust
trait T {}

struct A {}

impl T for A {}

struct B {}

impl T for B {}

#[allow(unreachable_code)]
fn new() -> impl T {
    #[cfg(feature = "feat")]
    return B {};

    return A {};
}

fn main() {
    let _ = new();
}
```

```
$ cargo clippy --all-features
    Checking analyser_bug v0.1.0 (/home/lenaic/Documents/dev/analyser_bug)
error[E0308]: mismatched types
  --> src/main.rs:16:12
   |
12 | fn new() -> impl T {
   |             ------ expected `B` because of return type
...
16 |     return A {};
   |            ^^^^ expected `B`, found `A`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `analyser_bug` (bin "analyser_bug") due to 1 previous error
```

### Version

```text
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: x86_64-unknown-linux-gnu
release: 1.76.0
LLVM version: 17.0.6
```

**Comment by wiiznokes** [comment]
Workarround:

put `#[cfg(not(feature = "feat"))]` on the top of `return A {};`

**Comment by y21** [comment]
This doesn't look like a bug. `impl T` always has to resolve to exactly one hidden type.

When you use `--all-features`, there will be two `return` statements with different types.
The first `return B {};` will resolve the `impl T` to be `B`, which then correctly reports a type-check if you later try to `return A {};`.

You should also see this error with `cargo check --feature feat`.

**Comment by wiiznokes** [comment]
I agree, it's more a rustc strange behavior. I'm going to close then
